### ISSUE TITLE
Use `$_SERVER` rather than `getenv()`

### ISSUE THREAD

**Issue by michaellenaghan** [marcantondahmen/automad#54]
When I moved from my development machine to my production host, `index.php` files started appearing in the URLs.

Using Automad's excellent debugging capabilities, I tracked the problem down to [here](https://github.com/marcantondahmen/automad/blob/0255fad39ac3a534a97164521b423839fa82a2bb/automad/src/Core/Config.php#L80). 

My host doesn't set a `SERVER_SOFTWARE` environment variable.

In fact, it turns out that using environment variables for such things is wrong.

I found an interesting 2016 thread [here](https://github.com/symphonycms/symphonycms/issues/2569). The long and short of it is that Symfony was using `getenv()` for certain kinds of things, and no longer does. (To be clear: it no longer uses `getenv()` for *server request vars*; it does for other things not related to server request vars.) [Laraval](https://github.com/laravel/framework) is the same.

It seems that the vars listed on the [`$_SERVER`](https://www.php.net/manual/en/reserved.variables.server.php) page should only be accessed through `$_SERVER`, and never through `getenv()`?

In the meantime, my workaround was to add this to `index.php`:

```
putenv('HTTP_HOST='.$_SERVER['HTTP_HOST']);
putenv('HTTP_X_FORWARDED_HOST='.$_SERVER['HTTP_X_FORWARDED_HOST']);
putenv('HTTP_X_FORWARDED_SERVER='.$_SERVER['HTTP_X_FORWARDED_SERVER']);
putenv('HTTPS='.$_SERVER['HTTPS']);
putenv('SERVER_NAME='.$_SERVER['SERVER_NAME']);
putenv('SERVER_SOFTWARE='.$_SERVER['SERVER_SOFTWARE']);
putenv('SCRIPT_NAME='.$_SERVER['SCRIPT_NAME']);
```

**Comment by michaellenaghan** [comment]
There are 23 uses of `getenv()` in 6 files. I checked them against the keys listed on the [`$_SERVER`](https://www.php.net/manual/en/reserved.variables.server.php) page; it looks like all of them should be switched to `$_SERVER`. (Btw, a couple of keys, like `SERVER_NAME`, are currently accessed both ways.)

Let me know if you'd like me to tackle this.

(Btw, just for completeness, my original list missed one key: `SERVER_PORT`.)

**Comment by marcantondahmen** [comment]
Hi, thanks for reporting this. The advantage of `getenv` is that it doesn't require a check whether a key exists. In that sense one could argue it is even the better solution. But I understand your trouble and will take a look at it. 

However, there is an easy way to make the `index.php` disappear from the URL. In case the automatic detection fails, you can set the `AM_INDEX` constant to an empty string in the config.

**Comment by michaellenaghan** [comment]
> In that sense one could argue it is even the better solution. But I understand your trouble and will take a look at it.

Marc, this isn't "my trouble." It's about the wrong approach and the right approach. There's no sense in which one could argue that the wrong approach is the better solution!

I sent you a thread in which the Symfony developers admitted they were wrong, and changed their code. (There's a lot of detail in that thread. It's worth a careful read. I followed links and looked at the source code they were talking about.)

After finding that, I took the time to verify that both Laravel and Symfony exclusively use `$_SERVER` for server request vars today.

[Here](https://github.com/getgrav/grav/search?q=%24_SERVER) is Grav, another flat-file PHP-based CMS. Same thing.

[Here](https://github.com/getkirby/kirby/search?q=%24_SERVER) is Kirby, another flat-file PHP-based CMS. Same thing.

If you like the convenience of what `getenv()` provides you can create a small wrapper around `$_SERVER` that works the same way.

**Comment by poetaster** [comment]
> I haven't tested if the putenv solution above will work with ha-proxy / varnish & a cdn, but I'll try to get to it.

No, that is not sufficient. I'm not sure what is happening here, but I can varnishlog the POST requests. I checked to make sure that no form / session cookies are being unset (are not), but login does not work. It does with the same setup on a drupal install.

**Comment by marcantondahmen** [comment]
One other thing came into my mind. There is the `AM_BASE_PROXY` constant that can be defined in the config in order to address issues with URL rewriting for pages behind proxies. Maybe you could give that one a quick try.

**Comment by poetaster** [comment]
> One other thing came into my mind. There is the `AM_BASE_PROXY` constant that can be defined in the config in order to address issues with URL rewriting for pages behind proxies. Maybe you could give that one a quick try.

I'll try this later this evening. Thanks!

**Comment by poetaster** [comment]
> One other thing came into my mind. There is the `AM_BASE_PROXY` constant that can be defined in the config in order to address issues with URL rewriting for pages behind proxies. Maybe you could give that one a quick try.

Errr, what should it be set to? In my drupal configs I noted 127.0.0.1 ?

**Comment by poetaster** [comment]
> One other thing came into my mind. There is the `AM_BASE_PROXY` constant that can be defined in the config in order to address issues with URL rewriting for pages behind proxies. Maybe you could give that one a quick try.

Errr, what should it be set to? In my drupal configs I noted 127.0.0.1 ? that value does not work ;)

**Comment by marcantondahmen** [comment]
After taking a closer look, it turned out that it needed more than changing from `getenv()` to `$_SERVER`. That actually has really nothing todo with it at all, as I mentioned before. However, I agree that a streamlined use of `$_SERVER` is more consistent here as well.

The actual issue is that the automatic determination of the base URL is not working correctly when running behind a proxy. That basically breaks the app. In the beginning it was not meant to run in such a scenario and also was not really tested therefore. 

However, I have just added the ability to run behind a proxy. The feature will be included in the next major release.

**Comment by poetaster** [comment]
> However, I have just added the ability to run behind a proxy. The feature will be included in the next major release.

Ah, that's great news. I host a lot of smaller sites on IPv6 only virts and proxy IPv4 with a transparent cache, so that would make those scenarios automadic :) Thanks for all your work!
